cylc/flow/commands.py

```diff
@@ -0,0 +1,434 @@
+from cylc.flow.workflow_status import RunMode, StopMode
+
+from metomi.isodatetime.parsers import TimePointParser
+
+if TYPE_CHECKING:
+    from cylc.flow.scheduler import Scheduler
+
+    # define a type for command implementations
+    Command = Callable[
+        ...,
+        AsyncGenerator,
+    ]
+
+# a directory of registered commands (populated on module import)
+COMMANDS: 'Dict[str, Command]' = {}
+
+
+def _command(name: str):
+    """Decorator to register a command."""
+    def _command(fcn: 'Command'):
+        nonlocal name
+        COMMANDS[name] = fcn
+        fcn.command_name = name  # type: ignore
+        return fcn
+    return _command
+
+
+async def run_cmd(fcn, *args, **kwargs):
+    """Run a command outside of the scheduler's main loop.
+
+    Normally commands are run via the Scheduler's command_queue (which is
+    correct), however, there are some use cases for running commands outside of
+    the loop:
+
+    * Running synchronous commands within the scheduler itself (e.g. on
+      shutdown).
+    * Integration tests (where you may want to cut out the main loop and
+      command queueing mechanism for simplicity).
+
+    For these purposes use "run_cmd", otherwise, queue commands via the
+    scheduler as normal.
+    """
+    cmd = fcn(*args, **kwargs)
+    await cmd.__anext__()  # validate
+    with suppress(StopAsyncIteration):
+        return await cmd.__anext__()  # run
+
+
+@_command('set')
+async def set_prereqs_and_outputs(
+    schd: 'Scheduler',
+    tasks: List[str],
+    flow: List[str],
+    outputs: Optional[List[str]] = None,
+    prerequisites: Optional[List[str]] = None,
+    flow_wait: bool = False,
+    flow_descr: Optional[str] = None,
+) -> AsyncGenerator:
+    """Force spawn task successors.
+
+    Note, the "outputs" and "prerequisites" arguments might not be
+    populated in the mutation arguments so must provide defaults here.
+    """
+    validate.consistency(outputs, prerequisites)
+    outputs = validate.outputs(outputs)
+    prerequisites = validate.prereqs(prerequisites)
+    validate.flow_opts(flow, flow_wait)
+
+    yield
+
+    if outputs is None:
+        outputs = []
+    if prerequisites is None:
+        prerequisites = []
+    yield schd.pool.set_prereqs_and_outputs(
+        tasks,
+        outputs,
+        prerequisites,
+        flow,
+        flow_wait,
+        flow_descr,
+    )
+
+
+@_command('stop')
+async def stop(
+    schd: 'Scheduler',
+    mode: Union[str, 'StopMode'],
+    cycle_point: Optional[str] = None,
+    # NOTE clock_time YYYY/MM/DD-HH:mm back-compat removed
+    clock_time: Optional[str] = None,
+    task: Optional[str] = None,
+    flow_num: Optional[int] = None,
+):
+    yield
+    if flow_num:
+        schd.pool.stop_flow(flow_num)
+        return
+
+    if cycle_point is not None:
+        # schedule shutdown after tasks pass provided cycle point
+        point = TaskID.get_standardised_point(cycle_point)
```

Comment:
For future work, we move code like this before the first `yield` in order to flag input errors at command queue time.